File: scripts/build-openapi.py
#!/usr/bin/env python3
"""
MileApp OpenAPI Build Script

Mimics Laravel ResourceController compilation logic.
Merges modular JSON files into single OpenAPI specs.

Usage:
    python scripts/build-openapi.py           # Build both
    python scripts/build-openapi.py public    # Build public only
    python scripts/build-openapi.py internal  # Build internal only
"""

import json
import yaml
import sys
from pathlib import Path

# Paths
PROJECT_ROOT = Path(__file__).parent.parent
API_SPECS = PROJECT_ROOT / "api-specs"
BASE_DIR = API_SPECS / "base"
PUBLIC_DIR = API_SPECS / "public"
INTERNAL_DIR = API_SPECS / "internal"
MODELS_DIR = API_SPECS / "Models"
OUTPUT_DIR = PROJECT_ROOT


def load_yaml(path):
    with open(path, 'r', encoding='utf-8') as f:
        return yaml.safe_load(f) or {}


def load_json(path):
    with open(path, 'r', encoding='utf-8') as f:
        # strict=False allows control characters in strings (like newlines)
        return json.load(f, strict=False)


def save_json(path, data):
    with open(path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
    print(f"✓ Generated: {path}")


def get_submodule_name(file_path):
    """Extract submodule: '1_Task.json' -> 'Task.json'"""
    name = file_path.name
    parts = name.split('_', 1)
    return parts[1] if len(parts) > 1 and parts[0].isdigit() else name


def merge_spec(base, module):
    """Merge module into base (paths, definitions, components, webhooks)"""
    if 'paths' in module and module['paths']:
        base.setdefault('paths', {}).update(module['paths'])
    if 'definitions' in module and module['definitions']:
        base.setdefault('definitions', {}).update(module['definitions'])
    if 'components' in module and module.get('components', {}).get('schemas'):
        base.setdefault('components', {}).setdefault('schemas', {}).update(module['components']['schemas'])
    if 'x-webhooks' in module and module['x-webhooks']:
        base.setdefault('x-webhooks', {}).update(module['x-webhooks'])


def build_public():
    """Build public API: base + public modules"""
    print("\n📦 Building PUBLIC API...")

    # Load base
    result = load_yaml(BASE_DIR / "initial.yaml") if (BASE_DIR / "initial.yaml").exists() else {}
    # Ensure paths/definitions are dicts (not None)
    result['paths'] = result.get('paths') or {}
    result['definitions'] = result.get('definitions') or {}

    # Load params
    if (BASE_DIR / "params.yaml").exists():
        result['parameters'] = load_yaml(BASE_DIR / "params.yaml")

    # Merge public modules
    for module_dir in sorted(PUBLIC_DIR.iterdir()) if PUBLIC_DIR.exists() else []:
        if module_dir.is_dir():
            for json_file in sorted(module_dir.glob("*.json")):
                merge_spec(result, load_json(json_file))
                print(f"  + {module_dir.name}/{json_file.name}")

    # Merge Models
    for json_file in sorted(MODELS_DIR.glob("*.json")) if MODELS_DIR.exists() else []:
        merge_spec(result, load_json(json_file))
        print(f"  + Models/{json_file.name}")

    return result


def build_internal():
    """Build internal API: base + internal (override) + public (fill gaps)"""
    print("\n🔒 Building INTERNAL API...")

    # Load base
    base_file = BASE_DIR / "initial_internal.yaml"
    if not base_file.exists():
        base_file = BASE_DIR / "initial.yaml"
    result = load_yaml(base_file) if base_file.exists() else {}
    # Ensure paths/definitions are dicts (not None)
    result['paths'] = result.get('paths') or {}
    result['definitions'] = result.get('definitions') or {}

    # Load params
    if (BASE_DIR / "params.yaml").exists():
        result['parameters'] = load_yaml(BASE_DIR / "params.yaml")

    # Track registered submodules (internal takes precedence)
    registered = {}

    # 1. Process INTERNAL first
    for module_dir in sorted(INTERNAL_DIR.iterdir()) if INTERNAL_DIR.exists() else []:
        if module_dir.is_dir():
            registered.setdefault(module_dir.name, {})
            for json_file in sorted(module_dir.glob("*.json")):
                submodule = get_submodule_name(json_file)
                merge_spec(result, load_json(json_file))
                registered[module_dir.name][submodule] = True
                print(f"  + (internal) {module_dir.name}/{json_file.name}")

    # 2. Process PUBLIC (skip if internal exists)
    for module_dir in sorted(PUBLIC_DIR.iterdir()) if PUBLIC_DIR.exists() else []:
        if module_dir.is_dir():
            for json_file in sorted(module_dir.glob("*.json")):
                submodule = get_submodule_name(json_file)
                if registered.get(module_dir.name, {}).get(submodule):
                    print(f"  - (skip) {module_dir.name}/{json_file.name}")
                    continue
                merge_spec(result, load_json(json_file))
                print(f"  + (public) {module_dir.name}/{json_file.name}")

    # Merge Models
    for json_file in sorted(MODELS_DIR.glob("*.json")) if MODELS_DIR.exists() else []:
        merge_spec(result, load_json(json_file))
        print(f"  + Models/{json_file.name}")

    return result


def main():
    print("=" * 50)
    print("MileApp OpenAPI Build")
    print("=" * 50)

    target = sys.argv[1] if len(sys.argv) > 1 else "all"

    if target in ["all", "public"]:
        spec = build_public()
        save_json(OUTPUT_DIR / "openapi-public.json", spec)
        print(f"  Paths: {len(spec.get('paths', {}))}")

    if target in ["all", "internal"]:
        spec = build_internal()
        save_json(OUTPUT_DIR / "openapi-internal.json", spec)
        print(f"  Paths: {len(spec.get('paths', {}))}")

    print("\n✅ Build complete!")


if __name__ == "__main__":
    main()

File: scripts/build-openapi.js
#!/usr/bin/env node
/**
 * MileApp OpenAPI Build Script
 *
 * Mimics Laravel ResourceController compilation logic.
 * Merges modular JSON files into single OpenAPI specs.
 *
 * Usage:
 *   node scripts/build-openapi.js           # Build both
 *   node scripts/build-openapi.js public    # Build public only
 *   node scripts/build-openapi.js internal  # Build internal only
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

// Paths
const PROJECT_ROOT = path.dirname(__dirname);
const API_SPECS = path.join(PROJECT_ROOT, 'api-specs');
const BASE_DIR = path.join(API_SPECS, 'base');
const PUBLIC_DIR = path.join(API_SPECS, 'public');
const INTERNAL_DIR = path.join(API_SPECS, 'internal');
const MODELS_DIR = path.join(API_SPECS, 'Models');
const OUTPUT_DIR = PROJECT_ROOT;

// Helper functions
function loadYaml(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    return yaml.load(content) || {};
  } catch (e) {
    return {};
  }
}

function loadJson(filePath) {
  try {
    let content = fs.readFileSync(filePath, 'utf8');
    // Fix unescaped control characters inside JSON string values
    // Laravel's JSON files have literal newlines in description strings
    content = content.replace(/"([^"\\]|\\.)*"/g, (match) => {
      return match
        .replace(/\n/g, '\\n')
        .replace(/\r/g, '\\r')
        .replace(/\t/g, '\\t');
    });
    return JSON.parse(content);
  } catch (e) {
    console.error(`  ⚠ Error loading ${filePath}: ${e.message}`);
    return {};
  }
}

function saveJson(filePath, data) {
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf8');
  console.log(`✓ Generated: ${filePath}`);
}

function getSubmoduleName(fileName) {
  // Extract submodule: '1_Task.json' -> 'Task.json'
  const parts = fileName.split('_');
  if (parts.length > 1 && /^\d+$/.test(parts[0])) {
    return parts.slice(1).join('_');
  }
  return fileName;
}

function getSortedDirs(dirPath) {
  if (!fs.existsSync(dirPath)) return [];
  return fs.readdirSync(dirPath)
    .filter(f => fs.statSync(path.join(dirPath, f)).isDirectory())
    .sort();
}

function getJsonFiles(dirPath) {
  if (!fs.existsSync(dirPath)) return [];
  return fs.readdirSync(dirPath)
    .filter(f => f.endsWith('.json'))
    .sort();
}

function mergeSpec(base, module) {
  // Merge paths
  if (module.paths && Object.keys(module.paths).length > 0) {
    base.paths = base.paths || {};
    Object.assign(base.paths, module.paths);
  }
  // Merge definitions (Swagger 2.0)
  if (module.definitions && Object.keys(module.definitions).length > 0) {
    base.definitions = base.definitions || {};
    Object.assign(base.definitions, module.definitions);
  }
  // Merge components.schemas (OpenAPI 3.0)
  if (module.components?.schemas && Object.keys(module.components.schemas).length > 0) {
    base.components = base.components || {};
    base.components.schemas = base.components.schemas || {};
    Object.assign(base.components.schemas, module.components.schemas);
  }
  // Merge x-webhooks
  if (module['x-webhooks'] && Object.keys(module['x-webhooks']).length > 0) {
    base['x-webhooks'] = base['x-webhooks'] || {};
    Object.assign(base['x-webhooks'], module['x-webhooks']);
  }
}

function buildPublic() {
  console.log('\n📦 Building PUBLIC API...');

  // Load base
  const initialPath = path.join(BASE_DIR, 'initial.yaml');
  let result = fs.existsSync(initialPath) ? loadYaml(initialPath) : {};

  // Ensure paths/definitions are objects (not null)
  result.paths = result.paths || {};
  result.definitions = result.definitions || {};

  // Load params
  const paramsPath = path.join(BASE_DIR, 'params.yaml');
  if (fs.existsSync(paramsPath)) {
    result.parameters = loadYaml(paramsPath);
  }

  // Merge public modules
  for (const moduleDir of getSortedDirs(PUBLIC_DIR)) {
    const modulePath = path.join(PUBLIC_DIR, moduleDir);
    for (const jsonFile of getJsonFiles(modulePath)) {
      const filePath = path.join(modulePath, jsonFile);
      const module = loadJson(filePath);
      mergeSpec(result, module);
      console.log(`  + ${moduleDir}/${jsonFile}`);
    }
  }

  // Merge Models
  for (const jsonFile of getJsonFiles(MODELS_DIR)) {
    const filePath = path.join(MODELS_DIR, jsonFile);
    const module = loadJson(filePath);
    mergeSpec(result, module);
    console.log(`  + Models/${jsonFile}`);
  }

  return result;
}

function buildInternal() {
  console.log('\n🔒 Building INTERNAL API...');

  // Load base
  let basePath = path.join(BASE_DIR, 'initial_internal.yaml');
  if (!fs.existsSync(basePath)) {
    basePath = path.join(BASE_DIR, 'initial.yaml');
  }
  let result = fs.existsSync(basePath) ? loadYaml(basePath) : {};

  // Ensure paths/definitions are objects (not null)
  result.paths = result.paths || {};
  result.definitions = result.definitions || {};

  // Load params
  const paramsPath = path.join(BASE_DIR, 'params.yaml');
  if (fs.existsSync(paramsPath)) {
    result.parameters = loadYaml(paramsPath);
  }

  // Track registered submodules (internal takes precedence)
  const registered = {};

  // 1. Process INTERNAL first
  for (const moduleDir of getSortedDirs(INTERNAL_DIR)) {
    registered[moduleDir] = registered[moduleDir] || {};
    const modulePath = path.join(INTERNAL_DIR, moduleDir);

    for (const jsonFile of getJsonFiles(modulePath)) {
      const submodule = getSubmoduleName(jsonFile);
      const filePath = path.join(modulePath, jsonFile);
      const module = loadJson(filePath);
      mergeSpec(result, module);
      registered[moduleDir][submodule] = true;
      console.log(`  + (internal) ${moduleDir}/${jsonFile}`);
    }
  }

  // 2. Process PUBLIC (skip if internal exists)
  for (const moduleDir of getSortedDirs(PUBLIC_DIR)) {
    const modulePath = path.join(PUBLIC_DIR, moduleDir);

    for (const jsonFile of getJsonFiles(modulePath)) {
      const submodule = getSubmoduleName(jsonFile);

      // Skip if internal version exists
      if (registered[moduleDir]?.[submodule]) {
        console.log(`  - (skip) ${moduleDir}/${jsonFile}`);
        continue;
      }

      const filePath = path.join(modulePath, jsonFile);
      const module = loadJson(filePath);
      mergeSpec(result, module);
      console.log(`  + (public) ${moduleDir}/${jsonFile}`);
    }
  }

  // Merge Models
  for (const jsonFile of getJsonFiles(MODELS_DIR)) {
    const filePath = path.join(MODELS_DIR, jsonFile);
    const module = loadJson(filePath);
    mergeSpec(result, module);
    console.log(`  + Models/${jsonFile}`);
  }

  return result;
}

function main() {
  console.log('='.repeat(50));
  console.log('MileApp OpenAPI Build');
  console.log('='.repeat(50));

  const target = process.argv[2] || 'all';

  if (target === 'all' || target === 'public') {
    const spec = buildPublic();
    saveJson(path.join(OUTPUT_DIR, 'openapi-public.json'), spec);
    console.log(`  Paths: ${Object.keys(spec.paths || {}).length}`);
  }

  if (target === 'all' || target === 'internal') {
    const spec = buildInternal();
    saveJson(path.join(OUTPUT_DIR, 'openapi-internal.json'), spec);
    console.log(`  Paths: ${Object.keys(spec.paths || {}).length}`);
  }

  console.log('\n✅ Build complete!');
}

main();
